dds/DCPS/DataReaderImpl.cpp

```diff
@@ -572,23 +572,19 @@ DataReaderImpl::remove_associations(const WriterIdSeq& writers,
 }
 
 void
-DataReaderImpl::remove_or_reschedule(const PublicationId& pub_id)
+DataReaderImpl::remove_publication(const PublicationId& pub_id)
 {
-  ACE_WRITE_GUARD(ACE_RW_Thread_Mutex, write_guard, this->writers_lock_);
-  WriterMapType::iterator where = writers_.find(pub_id);
-  if (writers_.end() != where) {
-    WriterInfo& info = *where->second;
-    WriterIdSeq writers;
-    push_back(writers, pub_id);
-    bool notify = info.notify_lost_;
-    if (info.removal_deadline_ < ACE_OS::gettimeofday()) {
-      write_guard.release();
-      remove_associations_i(writers, notify);
-    } else {
-      write_guard.release();
-      remove_associations(writers, notify);
+  WriterIdSeq writers;
+  bool notify = false;
+  {
+    ACE_WRITE_GUARD(ACE_RW_Thread_Mutex, write_guard, this->writers_lock_);
+    WriterMapType::iterator where = writers_.find(pub_id);
+    if (writers_.end() != where) {
```

Comment:
It would be nice to know that it works without knowing details of remove_associations_i(), but it's fine to leave it the way you have it.  We were discussing multiple exit points on another PR.  I still assert that multiple exit points are desirable in many cases, we have the RAII guard for a reason.